src/Init/Lean/PrettyPrinter/Parenthesizer.lean

```diff
@@ -0,0 +1,444 @@
+/-
+Copyright (c) 2020 Sebastian Ullrich. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+-/
+
+/-!
+The parenthesizer inserts parentheses into a `Syntax` object where syntactically necessary, usually as an intermediary
+step between the delaborator and the formatter. While the delaborator outputs structurally well-formed syntax trees that
+can be re-elaborated without post-processing, this tree structure is lost in the formatter and thus needs to be
+preserved by proper insertion of parentheses.
+
+# The abstract problem & solution
+
+The Lean 4 grammar is unstructured and extensible with arbitrary new parsers, so in general it is undecidable whether
+parentheses are necessary or even allowed at any point in the syntax tree. Parentheses for different categories, e.g.
+terms and levels, might not even have the same structure. In this module, we focus on the correct parenthesization of
+parsers defined via `Lean.Parser.prattParser`, which includes both aforementioned built-in categories. Custom
+parenthesizers can be added for new node kinds, but the data collected in the implementation below might not be
+appropriate for other parenthesization strategies.
+
+Usages of a parser defined via `prattParser` in general have the form `p rbp`, where `rbp` is the right-binding power.
+Recall that a Pratt parser greedily parses a leading token with precedence at least `rbp` (otherwise it fails) followed
+by zero or more trailing tokens with precedence *higher* than `rbp`. Thus we should parenthesize a syntax node `stx`
```

Comment:
@leodemoura I may have asked this before, but could we get rid of this difference in precedence handling between leading and trailing tokens? It would simplify both documentation and code.

Reply:
I don't remember the proposal. It takes me a long time to remember all the details, and it is always a traumatic experience. (e.g., a lot of stuff breaks). I recall we revised the precedence handling a few months ago, and we all seemed happy.